### PR TITLE
New footer design implementation and rewrite of FooterButton class

### LINKED ISSUE #5923: Small overlap between buttons at the bottom of song select
**Describe the bug:**

There is a small (1 pixel?) overlap between the buttons at the bottom of song select.

**Screenshots or videos showing encountered issue:** 

image

**osu!lazer version:** 2019.831.0

### PR DESCRIPTION
**Pull request by mk56-spn**
- Addresses https://github.com/ppy/osu/discussions/20135#
- Resolves https://github.com/ppy/osu/issues/5923

Tried to stick as best as i could to the designs

matched the animations and such with ShearedButton.cs for consistency and rewrote the code to more closely track with that.

Sample Vid: 

https://user-images.githubusercontent.com/74463310/188483167-e0475761-9ec4-40c3-8cf4-0e85faf0e8a2.mov

will do the back button and Logo replacement seperately since they are more finnicky and lots of screens depend on them

### CODE REVIEW

**Comment by peppy** [comment]
The metrics seem really off from the design. Is there a reason for that?

**Comment by mk56-spn** [comment]
> The metrics seem really off from the design. Is there a reason for that?

 some i changed purposefully because they didn't match the sized you would expect from the design if i took the values at face value, especially the text , the base button width, etc.. 
 
 i tested at 1x scale on a 1080 p screen and both are in pixels so im not sure why it could be. if you want i can match them exactly

**Comment by Theighlin** [comment]
@mk56-spn have you read #17940?

**Comment by peppy** [comment]
> some i changed purposefully because they didn't match the sized you would expect from the design if i took the values at face value, especially the text , the base button width, etc..

I'd expect them to match visually if that's the case, but they don't.

File: osu.Game/Screens/Select/FooterButtonMods.cs
Comment (Joehuu) [inline osu.Game/Screens/Select/FooterButtonMods.cs:65]:
```suggestion
            IconUsageBox = FontAwesome.Solid.ExchangeAlt;
```

File: osu.Game/Screens/Select/FooterButtonRandom.cs
Comment (Joehuu) [inline osu.Game/Screens/Select/FooterButtonRandom.cs:33]:
```suggestion
            IconUsageBox = FontAwesome.Solid.Random;
```

**Review by Joehuu — COMMENTED** [review]
FA 6 renames a lot of the icons. The links I gave in the discussion show aliases from 5.

File: osu.Game/Screens/Select/FooterButtonMods.cs
Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButtonMods.cs]: done

File: osu.Game/Screens/Select/FooterButtonRandom.cs
Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButtonRandom.cs]: done

**Comment by mk56-spn** [comment]
> > some i changed purposefully because they didn't match the sized you would expect from the design if i took the values at face value, especially the text , the base button width, etc..
> 
> I'd expect them to match visually if that's the case, but they don't.

i apparently had the shear set up incorrectly which was causing a lot of the issues i saw when working on it

one thing, ive resized the bar because its 60 px tall in this iteration of design are you ok with this  looking as such:
unaligned
shall  i resize back to 50 for now or just make sure to get the back button pr done before this needs merging?

**Comment by Stedoss** [comment]
Due to the refactor to `FooterButton`, the `Rewind` action (right click on the `FooterButtonRandom` button) behaves a little strange in relation to animations. For example, right clicking on the button will not call the base `OnMouseUp()` function and won't scale back to the original size because of this early return: https://github.com/ppy/osu/blob/97e85ef0d844a2de23715b547dde2bf89713af44/osu.Game/Screens/Select/FooterButtonRandom.cs#L121-L131

Also of note, it seems that in the `FooterButtonRandom` class, the `OnPressed()` and `OnReleased()` don't call their base implementations, which means that using F2 to trigger the random button does not play the scaling animations.

File: osu.Game/Screens/Select/FooterButton.cs
Comment (peppy) [inline osu.Game/Screens/Select/FooterButton.cs:32]: Where did this colour come from?

File: osu.Game/Screens/Select/FooterButton.cs
Comment (peppy) [inline osu.Game/Screens/Select/FooterButton.cs:79]:
```suggestion
                Offset = new Vector2(0.25f, 2),
                Colour = Colour4.Black.Opacity(0.25f),
```

Comment (peppy) [reply osu.Game/Screens/Select/FooterButton.cs:79]: Also interested to know where these values came from. Just visual ballparking?

File: osu.Game/Screens/Select/FooterButton.cs
Comment (peppy) [inline osu.Game/Screens/Select/FooterButton.cs:97]: ??

File: osu.Game/Screens/Select/FooterButton.cs
Comment (peppy) [inline osu.Game/Screens/Select/FooterButton.cs:131]: No number should ever be this precise. I don't know how you sourced this but it's not right.

File: osu.Game/Screens/Select/FooterButton.cs
Comment (peppy) [inline osu.Game/Screens/Select/FooterButton.cs:30]: Move inline.

File: osu.Game/Screens/Select/FooterButton.cs
Comment (peppy) [inline osu.Game/Screens/Select/FooterButton.cs:160]: Again, this is way too (incorrectly) precise.

File: osu.Game/Screens/Select/FooterButton.cs
Comment (peppy) [inline osu.Game/Screens/Select/FooterButton.cs:130]: Is there a reason this `SHEAR` de-application is done in so many places rather than a single container?

File: osu.Game/Screens/Select/FooterButtonMods.cs
Comment (peppy) [inline osu.Game/Screens/Select/FooterButtonMods.cs:42]: This can't be in 100 places.

**Review by peppy — CHANGES_REQUESTED** [review]
In addition to other things pointed out, just an initial pass.

**Comment by peppy** [comment]
The tweening and animation style is completely wrong with this change, but I can fix that once the other issues are fixed. Maybe. For now, please remove the scaling of the buttons during mouse down completely. And change the buttons to be anchored to the bottom of the parent, not the top.

@arflyte please advise how you imagined mods showing in the footer, if at all. this ad-hoc implementation looks very bad:

https://user-images.githubusercontent.com/191335/188546146-f0956449-3d53-47d9-a2b5-ffb64e51bbee.mp4

File: osu.Game/Screens/Select/FooterButton.cs
Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButton.cs:131]: Figma

File: osu.Game/Screens/Select/FooterButton.cs
Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButton.cs:32]: As with all the others I sourced them straight from figma.

Comment (peppy) [reply osu.Game/Screens/Select/FooterButton.cs:32]: And this colour doesn't match an existing scheme? Can you link or screenshot where you're seeing this?

Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButton.cs:32]: sorry i should've looked for matching schemes, will address, but still linking this as proof i didn't magic the colour out of nowhere,
sourced colour

File: osu.Game/Screens/Select/FooterButton.cs
Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButton.cs:160]: size
shall i go with 120 then?

File: osu.Game/Screens/Select/FooterButton.cs
Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButton.cs:97]: i had issues with keeping the container sized appropriately whilst allowing for the expansion with mods (which was a misguided attempt to preserve the previous functionality, my reasoning for leaving it in) will be removed with the rest

File: osu.Game/Screens/Select/FooterButton.cs
Comment (smoogipoo) [reply osu.Game/Screens/Select/FooterButton.cs:32]: I'm fairly certain the `B/B3` in the `Colours` section of figma indicates that this is a [Background3](https://github.com/ppy/osu/blob/6296c971f8d7a8820b3f40162d33b7d219b7d0c9/osu.Game/Overlays/OverlayColourProvider.cs#L47) which is pretty much the same as the `Aquamarine` scheme (#394641 vs #394642)

File: osu.Game/Screens/Select/FooterButton.cs
Comment (peppy) [reply osu.Game/Screens/Select/FooterButton.cs:131]: @arflyte can you check on this?

File: osu.Game/Screens/Select/FooterButton.cs
Comment (peppy) [reply osu.Game/Screens/Select/FooterButton.cs:160]: I'd probably have made it relative X + padding, else it won't expand properly if the button is made to grow wider.

**Comment by mk56-spn** [comment]
> The tweening and animation style is completely wrong with this change, but I can fix that once the other issues are fixed. Maybe. For now, please remove the scaling of the buttons during mouse down completely. And change the buttons to be anchored to the bottom of the parent, not the top.
> 
> @arflyte please advise how you imagined mods showing in the footer, if at all. this ad-hoc implementation looks very bad:
> 
> https://user-images.githubusercontent.com/191335/188546146-f0956449-3d53-47d9-a2b5-ffb64e51bbee.mp4
> 
> 

Waiting on the answer to this since I've restructured the nesting hierarchy for drawables and I need to know if I can do away with the mods box before I continue

File: osu.Game/Screens/Select/FooterButton.cs
Comment (arflyte) [reply osu.Game/Screens/Select/FooterButton.cs:129]: I check on my side and it says 40, not some floating number.

**Comment by arflyte** [comment]
That is definitely not the way to display selected mods. There's supposed to be a floating indicator on top of the button, but I haven't designed that yet.

Place the selected mod floating on top of the button for now, until I work on its fix.

File: osu.Game/Screens/Select/FooterButtonMods.cs
Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButtonMods.cs]: done

File: osu.Game/Screens/Select/FooterButton.cs
Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButton.cs]: adressed

File: osu.Game/Screens/Select/FooterButton.cs
Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButton.cs]: done

**Comment by mk56-spn** [comment]
> That is definitely not the way to display selected mods. There's supposed to be a floating indicator on top of the button, but I haven't designed that yet.
> 
> Place the selected mod floating on top of the button for now, until I work on its fix.

addressed 
testing

File: osu.Game/Screens/Select/FooterButton.cs
Comment (mk56-spn) [reply osu.Game/Screens/Select/FooterButton.cs]: done

**Comment by peppy** [comment]
osu! 2022-09-08 at 08 58 17

Centering looks wrong.

Also please restore the height of the footer back to what it was. Even though this is a temporary change, it looks worse with it being taller in only this location:

osu! 2022-09-08 at 08 59 34

osu! 2022-09-08 at 08 59 45

**Comment by mk56-spn** [comment]
> osu! 2022-09-08 at 08 58 17
> 
> Centering looks wrong.
> 
> Also please restore the height of the footer back to what it was. Even though this is a temporary change, it looks worse with it being taller in only this location:
> 
> osu! 2022-09-08 at 08 59 34
> 
> osu! 2022-09-08 at 08 59 45

Yeah im aware i didnt req a pr yet because im trying to figure out how to use the overlaycolourprovider to feed the colours yet, sorry i shouldve perhaps mentioned that

and yep i was gonna handle the centering today

**Comment by peppy** [comment]
Lol

osu! 2022-09-28 at 08 27 17

I don't even know what to do with this PR. I think it's probably blocked until @arflyte provides a proper design.

**Comment by mk56-spn** [comment]
> Lol
> 
> osu! 2022-09-28 at 08 27 17
> 
> I don't even know what to do with this PR. I think it's probably blocked until @arflyte provides a proper design.

well i was told to put it up there in my defense i agree that that overlap sucks,

**Comment by mk56-spn** [comment]
if possible id love a quick review of my latest commit, i think the structuring of the drawable is much better now and i removed all the margin jankiness i'd done due to lack of knowledge. not pressing at all since this is blocked for now regardless, just would be neat.
shear centering should also be better

**Comment by peppy** [comment]
@arflyte has there been any progress around this design?

**Comment by peppy** [comment]
@mk56-spn there's been some design work on the mod display it would seem (https://www.figma.com/file/DXKwqZhD5yyb1igc3mKo1P/Song-Select-Draft-2?node-id=0%3A1&t=bRPeY78djO0JkLzS-0, bottom left)

Safari 2022-11-23 at 09 24 40

That said, it's probably going to require a bit of implementation detail so I'd argue that the display should just be removed from this PR completely and added as a separate effort.

**Comment by peppy** [comment]
Needs non-trivial conflict resolution.

**Comment by mk56-spn** [comment]
> That said, it's probably going to require a bit of implementation detail so I'd argue that the display should just be removed from this PR completely and added as a separate effort.

Sounds good. Fwiw I'm willing to do this part as well. But yeah I'd rather get practice mode and the other bits I have in my pipeline out so I'll put this secondary effort on the backburner If thats ok

**Comment by mk56-spn** [comment]
not for review currently, adressing that rather horrid conflict

**Comment by peppy** [comment]
@arflyte still missing design



what do

**Comment by mk56-spn** [comment]
> @arflyte still missing design
> 
> osu! 2022-12-01 at 06 44 46
> 
> what do

i can make a semi fitting placeholder change for them.

in other words replace their current colours with B6/ B5 background colour, move the colour itself to a thin bar on one of the sides , add some corner radius rounding, a bit of spacing between them , make them less tall and shift them up so they dont collide with the new buttons.

**Comment by peppy** [comment]
I honestly think you should implement your new buttons in complete isolation. In a test scene. So you don't have to remove the multiplier display code and wait for all other design elements to be completed.

Trying to immediately slot this in amongst many other surrounding elements is, as we have seen from this PR, not a good idea.

You're welcome to suffix your component with a V2 as has been done elsewhere. Opening in a new PR is also fine if it cleans up the history.

**Comment by mk56-spn** [comment]
yep ive upped my commit practices since making this and i totally agree this PR looks rather filthy, gonna close this in a bit in favour of a clean branch

**Pull request closed without merging**
